zebra-rpc/src/methods.rs

```diff
@@ -613,24 +613,62 @@ where
                 // This RPC is used in `lightwalletd`'s initial sync of 2 million blocks,
                 // so it needs to load all its fields very efficiently.
                 //
-                // Currently, we get the transaction IDs from an index, which is much more
-                // efficient than loading all the block data and hashing all the transactions.
+                // Currently, we get the block hash and transaction IDs from indexes,
+                // which is much more efficient than loading all the block data,
+                // then hashing the block header and all the transactions.
 
-                // TODO: look up the hash if we only have a height,
-                //       and look up the height if we only have a hash
-                let hash = hash_or_height.hash().map(GetBlockHash);
-                let height = hash_or_height.height();
+                // Get the block hash from the height -> hash index, if needed
+                //
+                // # Concurrency
+                //
+                // For consistency, this lookup must be performed first, then all the other
+                // lookups must be based on the hash.
+                //
+                // All possible responses are valid, even if the best chain changes. Clients
+                // must be able to handle chain forks, including a hash for a block that is
+                // later discovered to be on a side chain.
+
+                let hash = match hash_or_height {
```

Comment:
Optional/Nitpick: use or remove the `HashOrHeight.hash()` method

Reply:
We can't use that method here, because this match statement is async and uses the `?` return operator.